README_cmake.md

```diff
@@ -304,17 +695,219 @@ and the `USE_RUNTIME` option is ignored. Other options work as expected.
+module.
+
+One would then use Halide to build a generator executable in one self-contained
+project, then export that target to be used in a separate project. The second
+project would be configured with the target toolchain and would call
+`add_halide_library` with no `TARGETS` option and set `FROM` equal to the name
+of the imported generator executable. Obviously, this is a significant increase
+in complexity over a typical CMake project.
+
+### Use ExternalProject directly
+
+A lighter weight alternative to the above is to use `ExternalProject` directly
+in your parent build. Configure the parent build with the target toolchain, and
+configure the inner project to use the host toolchain. Then, manually create an
+IMPORTED target for your generator executable and call `add_halide_library` as
+described above.
+
+The main drawback of this approach is that creating accurate IMPORTED targets is
+difficult since predicting the names and locations of your binaries across all
+possible platform and CMake project generators is difficult. In particular, it
+is hard to predict executable extensions in cross-OS builds.
+
+### Use an emulator or run on device
+
+The
+[`CMAKE_CROSSCOMPILING_EMULATOR`](https://cmake.org/cmake/help/latest/variable/CMAKE_CROSSCOMPILING_EMULATOR.html)
+variable allows one to specify a command _prefix_ to run a target-system binary
+on the host machine. One could set this to a custom shell script that uploads
+the generator executable, runs it on the device and copies back the results.
+
+### Bypass CMake
+
+The previous two options ensure that the targets generated by
+`add_halide_library` will be _normal_ static libraries. This approach does not
+use ExternalProject, but instead produces `IMPORTED` targets. The main drawback
+of `IMPORTED` targets is that they are considered second-class in CMake. In
+particular, they cannot be installed with the typical `install(TARGETS)`
+command. Instead, they must be installed using `install(FILES)` and the
+`$<TARGET_FILE:>` generator expression.
+
+# Contributing CMake code to Halide
+
+When contributing new CMake code to Halide, keep in mind that the minimum
+version is 3.16. Therefore, it is possible (and indeed required) to use modern
+CMake best practices.
+
+Like any large and complex system with a dedication to preserving backwards
+compatibility, CMake is difficult to learn and full of traps. While not
+comprehensive, the following serves as a guide for writing quality CMake code
+and outlines the code quality expectations we have as they apply to CMake.
+
+## General guidelines and best practices
+
+The following are some common mistakes that lead to subtly broken builds.
+
+- **Reading the build directory.** While setting up the build, the build
+  directory should be considered _write only_. Using the build directory as a
+  read/write temporary directory is acceptable as long as all temp files are
```

Comment:
Nit: why not just forbid read/write access entirely? Don't have to worry about cleaning up temp files if they aren't ever created...

Reply:
Writing is okay if you're doing `configure_file` or the like. It's more about avoiding a bug we used to have that read a file that was later created, so it would work on subsequent runs, but not the first.